streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala

```diff
@@ -1980,7 +1980,23 @@ class ZStream[-R, +E, +A](val channel: ZChannel[R, Any, Any, Any, E, Chunk[A], A
    * Takes the last specified number of elements from this stream.
    */
   def takeRight(n: Int): ZStream[R, E, A] =
-    ???
+    if (n <= 0) ZStream.empty
+    else
+      new ZStream(
+        ZChannel.unwrap(
+          for {
+            queue <- Queue.sliding[A](n)
+          } yield {
+            lazy val reader: ZChannel[Any, E, Chunk[A], Any, E, Chunk[A], Unit] = ZChannel.readWith(
+              (in: Chunk[A]) => ZChannel.fromEffect(queue.offerAll(in)) *> reader,
```

Comment:
Would you be alright with using a plain mutable Queue here? should be much faster.

Reply:
Would @adamgfraser 's RingBuffer work here as well?

Reply:
We need a `SingleThreadedRingBuffer` here. Let me do a quick PR to expose the functionality we want.

Reply:
Implemented in #4951.

Reply:
@paulpdaniels This is available on the `series/2.x` branch now.